### DISCUSSION TITLE
How to log experiments with permission

### DISCUSSION THREAD

**Discussion by fabiosanger** [mlflow-oidc/mlflow-oidc-auth#24] [General]
.bashrc

```
export MLFLOW_TRACKING_USERNAME='xxx
export MLFLOW_TRACKING_PASSWORD='xx
export MLFLOW_TRACKING_TOKEN='x'
export MLFLOW_OIDC_DISCOVERY_URL="https://xxx.okta.com/.well-known/openid-configuration"
export MLFLOW_OIDC_CLIENT_SECRET="xx-xxx"
export MLFLOW_OIDC_CLIENT_ID="xx"
```

TRACKING_SERVER_HOST points to the authenticated server

```
mlflow.set_tracking_uri(uri=f"http://{TRACKING_SERVER_HOST}/")
print("First Experiment Elastic Net")
permission = "EDIT"
experiment_name = f'exp_EL_{exp_version}_perm_{permission}'
exp = mlflow.set_experiment(experiment_name=experiment_name)
```
the last line returns

> File "/nfs/users/nfs_f/fg12/repos/mlflow-authentication/.env_mlops/lib/python3.11/site-packages/mlflow/utils/rest_utils.py", line 180, in verify_rest_response
>     raise MlflowException(
> mlflow.exceptions.MlflowException: API request to endpoint /api/2.0/mlflow/experiments/permissions/create failed with error code 401 != 200. Response body: 'You are not authenticated. Please see documentation for detailshttps://github.com/data-platform-hq/mlflow-oidc-auth'

**Comment by fabiosanger** [comment]
the initial issue has been fixed, the solution was to use the Access Key as password.

Now I am trying to set the permission on the experiment 

```
    mlflow.set_tracking_uri(uri=f"http://{TRACKING_SERVER_HOST}/")
    from mlflow.server.auth.client import AuthServiceClient
    # # Create an instance of the AuthServiceClient
    auth_client = AuthServiceClient(tracking_uri=mlflow.get_tracking_uri())
    exp_version = "auth_v1.2"
    ###########First Experiment Elastic Net #############
    print("First Experiment Elastic Net")
    permission = "EDIT"
    experiment_name = f'exp_EL_{exp_version}_perm_{permission}'
    try:
        # Attempt to create a new experiment
        exp_id = mlflow.create_experiment(name=experiment_name)
    except mlflow.exceptions.RestException as e:
        if "RESOURCE_ALREADY_EXISTS" in str(e):
            # If the experiment already exists, get the experiment ID
            exp_id = mlflow.get_experiment_by_name(experiment_name).experiment_id
        else:
            raise e
    # Get experiment details using the experiment ID
    exp = mlflow.get_experiment(exp_id)

    experiment_permission = auth_client.create_experiment_permission(
        experiment_id=exp_id,
        username=os.environ['MLFLOW_TRACKING_USERNAME'],
        permission=permission
    )
```

> Traceback (most recent call last):
>   File "/nfs/users/nfs_f/fg12/repos/mlflow-authentication/src/multi_exp_auth.py", line 123, in <module>
>     experiment_permission = auth_client.create_experiment_permission(
>                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/nfs/users/nfs_f/fg12/repos/mlflow-authentication/.env_mlops/lib/python3.11/site-packages/mlflow/server/auth/client.py", line 266, in create_experiment_permission
>     resp = self._request(
>            ^^^^^^^^^^^^^^
>   File "/nfs/users/nfs_f/fg12/repos/mlflow-authentication/.env_mlops/lib/python3.11/site-packages/mlflow/server/auth/client.py", line 37, in _request
>     resp = http_request_safe(host_creds, endpoint, method, **kwargs)
>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/nfs/users/nfs_f/fg12/repos/mlflow-authentication/.env_mlops/lib/python3.11/site-packages/mlflow/utils/rest_utils.py", line 170, in http_request_safe
>     return verify_rest_response(response, endpoint)
>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/nfs/users/nfs_f/fg12/repos/mlflow-authentication/.env_mlops/lib/python3.11/site-packages/mlflow/utils/rest_utils.py", line 178, in verify_rest_response
>     raise RestException(json.loads(response.text))
> mlflow.exceptions.RestException: INVALID_PARAMETER_VALUE: Missing value for required parameter 'user_name'. See the API docs for more information about request parameters.

**Comment by maxtenzin** [comment]
Hey @fabiosanger how did you manage to get the authentication working? I'm receiving `MlflowException: API request to endpoint /api/2.0/mlflow/runs/create failed with error code 401 != 200. Response body: 'You are not authenticated. Please see documentation for detailshttps://github.com/data-platform-hq/mlflow-oidc-auth'` always

**Comment by fabiosanger** [comment]
Just got back from AL, i will reply as soon as possible
